### PR TITLE
Fix delegate_facts with interpreter not being set

### LINKED ISSUE #70168: 2.9.10: UnboundLocalError: local variable 'module_style' referenced before assignment
##### SUMMARY

after upgrade from 2.9.9 to 2.9.10 we see an exception in a task using ~~lineinfile~~ `delegate_facts=true`

##### ISSUE TYPE
- Bug Report

##### ANSIBLE VERSION

```paste below
2.9.10
```

##### CONFIGURATION

```paste below
ANSIBLE_NOCOWS(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = True
ANSIBLE_PIPELINING(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = True
ANSIBLE_SSH_ARGS(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = -o ControlMaster=auto -o ControlPersist=60s -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no
CACHE_PLUGIN(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = jsonfile
CACHE_PLUGIN_CONNECTION(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = .cache/facts
CACHE_PLUGIN_TIMEOUT(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = 86400
COLLECTIONS_PATHS(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = [u'/home/rmoser/Projects/puzzle/ansible-puzzle/collections_galaxy']
CONDITIONAL_BARE_VARS(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = False
DEFAULT_BECOME(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = True
DEFAULT_BECOME_ASK_PASS(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = False
DEFAULT_BECOME_METHOD(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = sudo
DEFAULT_BECOME_USER(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = root
DEFAULT_FORKS(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = 20
DEFAULT_GATHERING(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = smart
DEFAULT_HOST_LIST(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = [u'/home/rmoser/Projects/puzzle/ansible-puzzle/inventories/cloud_staging']
DEFAULT_LOAD_CALLBACK_PLUGINS(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = False
DEFAULT_MANAGED_STR(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = Warning: File is managed by Ansible
DEFAULT_PRIVATE_KEY_FILE(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = /home/rmoser/.ssh/id_rsa
DEFAULT_REMOTE_USER(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = ansible
DEFAULT_ROLES_PATH(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = [u'/home/rmoser/Projects/puzzle/ansible-puzzle/roles_galaxy', u'/home/rmoser/Projects/puzzle/ansible-puzzle/roles']
DEFAULT_STDOUT_CALLBACK(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = yaml
HOST_KEY_CHECKING(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = True
INTERPRETER_PYTHON(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = auto
RETRY_FILES_ENABLED(/home/rmoser/Projects/puzzle/ansible-puzzle/ansible.cfg) = False

```

##### STEPS TO REPRODUCE

```yaml
---
- hosts: my.example.com
  gather_facts: yes
  tasks:
  - name: make something
    delegate_facts: True
    delegate_to: other.example.com
    lineinfile:
      path: /tmp/foo.txt
      regexp: '^Listen '
      insertafter: '^#Listen '
      line: Listen 8080

```

##### ACTUAL RESULTS

```paste below
The full traceback is:
Traceback (most recent call last):
  File "/home/rmoser/Projects/resmo/ansible/lib/ansible/executor/task_executor.py", line 146, in run
    res = self._execute()
  File "/home/rmoser/Projects/resmo/ansible/lib/ansible/executor/task_executor.py", line 654, in _execute
    result = self._handler.run(task_vars=variables)
  File "/home/rmoser/Projects/resmo/ansible/lib/ansible/plugins/action/normal.py", line 46, in run
    result = merge_hash(result, self._execute_module(task_vars=task_vars, wrap_async=wrap_async))
  File "/home/rmoser/Projects/resmo/ansible/lib/ansible/plugins/action/__init__.py", line 831, in _execute_module
    (module_style, shebang, module_data, module_path) = self._configure_module(module_name=module_name, module_args=module_args, task_vars=task_vars)
  File "/home/rmoser/Projects/resmo/ansible/lib/ansible/plugins/action/__init__.py", line 244, in _configure_module
    return (module_style, module_shebang, module_data, module_path)
UnboundLocalError: local variable 'module_style' referenced before assignment
fatal: [server.example.com]: FAILED! => 
  msg: Unexpected failure during module execution.
  stdout: ''

```

### PR DESCRIPTION
**Pull request by mkrizek**
##### SUMMARY

Fixes https://github.com/ansible/ansible/issues/70168

Introduced in https://github.com/ansible/ansible/pull/69604 which was backported to `2.9.10`. This would need to be backported to `2.9` and `2.10`.

This is another take to get these fixes in after they were reverted in https://github.com/ansible/ansible/pull/70272 due to issues with testing the fixes in our CI.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

` lib/ansible/plugins/action/__init__.py`

### CODE REVIEW

**Comment by mkrizek** [comment]
cc @mattclay @samdoran @bcoca

**Comment by mkrizek** [comment]
From https://github.com/ansible/ansible/pull/70171#discussion_r445135210:
>  The non_local entry should not be added to existing tests, since it disables coverage support for those tests. If it is needed for the test to work, a new test needs to be created.

@mattclay so this test needs to run under the following conditions:

1. task needs to be delegated (`delegate_to`)
2. task needs to delegate facts (`delegate_facts: yes`)
3. task does NOT set `ansible_python_interpreter` or sets it to `auto`. Since `ansible-test` forces tests to set `ansible_python_interpreter` variable I set it for this test purpose to `auto` which caused the `known_hosts_hash` error. I noticed that `interpreter_discovery_python` uses `non_local` to workaround that (at least it seemed so). Is there other way to achieve that without setting `non_local` or should I create a new target for this test?

File: lib/ansible/plugins/action/__init__.py [hunk 1]
```diff
@@ -242,18 +242,22 @@ def _configure_module(self, module_name, module_args, task_vars=None):
...
+                    use_vars['ansible_facts'] = {}
+                use_vars['ansible_facts'][discovered_key] = self._discovered_interpreter
+
                 # TODO: this condition prevents 'wrong host' from being updated
                 # but in future we would want to be able to update 'delegated host facts'
                 # irrespective of task settings
                 if not self._task.delegate_to or self._task.delegate_facts:
                     # store in local task_vars facts collection for the retry and any other usages in this worker
-                    if use_vars.get('ansible_facts') is None:
+                    if task_vars.get('ansible_facts') is None:
```
Comment (bcoca) [inline lib/ansible/plugins/action/__init__.py]: we should always have ansible_facts here, the if/set to empty should not be needed here, nor above.

**Review by bcoca — APPROVED** [review]

File: lib/ansible/plugins/action/__init__.py [hunk 2]
```diff
@@ -160,7 +160,7 @@ def _configure_module(self, module_name, module_args, task_vars=None):
         '''
 
         if task_vars is None:
```
Comment (bcoca) [inline lib/ansible/plugins/action/__init__.py]: not sure this is needed either, its only called from 2 locations and they always pass task_vars .. i would make it a required field at this point

not a blocker to this PR, just random note

**Review by mattclay — APPROVED** [review]

File: lib/ansible/plugins/action/__init__.py (hunk 2 above)
Comment (mkrizek) [reply lib/ansible/plugins/action/__init__.py]: I went ahead and made the `task_vars` required method argument in this PR.

**Pull request merged**